lib/common/ipc.c

```diff
@@ -278,6 +278,58 @@ crm_client_disconnect_all(qb_ipcs_service_t *service)
     }
 }
 
+/*!
+ * \internal
+ * \brief Allocate a new crm_client_t object based on an IPC connection
+ *
+ * \param[in] c           IPC connection (or NULL to allocate generic client)
+ * \param[in] key         Connection table key (or NULL to use sane default)
+ * \param[in] uid_client  UID corresponding to c (ignored if c is NULL)
+ *
+ * \return Pointer to new crm_client_t (or NULL on error)
+ */
+static crm_client_t *
+client_from_connection(qb_ipcs_connection_t *c, void *key, uid_t uid_client)
+{
+    crm_client_t *client = calloc(1, sizeof(crm_client_t));
+
+    if (client == NULL) {
+        crm_perror(LOG_ERR, "Allocating client");
+        return NULL;
+    }
+
+    if (c) {
+#if ENABLE_ACL
+        client->user = uid2username(uid_client);
+        if (client->user == NULL) {
+            client->user = strdup("#unprivileged");
```

Comment:
No justification for this choice/solution was provided.

Is it meant to allow for ad-hoc workarounds (since `acl_target` is unrestricted
value in the schema when for instance SELinux starts to prevents `/etc/passwd`
access for pacemaker daemons?

Reply:
Just that if we bump the major CIB version so as to sanitize a crazy
situations around ACLs vs. non-uniqueness/overkill uniqueness
(alternatively, retroactively since 3.0, but rather not), such
configuration quirk would stop working, for not matching `NCName`
and hence would get dropped.

Hence if such workaround shall be viable, there's not much wiggle
room than to use something possibly, yet practically impossibly
clashing, like `__PCMK-UNPRIVILEGED-USER__`
(per `passwd(5)`:

> It should not contain capital letters.

)

Reply:
You are stuck with single and only scenario in mind (yeah, nice find,
nice fixes), but won't generalize on other failure modes, e.g. mentioned
SELinux. One of the apparent solutions if that non-existent user is
observable somewhere in the log messages, would be to put node
into maintenance, stop pacemaker, force-feed it with new bumped
CIB containing this new ACL entry relating to a role already defined,
starting it back, unmaintaining, letting the new config spread (perhaps
oversimplified), problem temporarily solved.

When someone can mangle with the users on the system (or indirectly
through separate user namespace/containerization, chances are lost
already), so I would at least legitimize such extreme workaround and
have it persist said `NCName` unification in the schema, that the current
value, when configured, would not survive.  Btw. not with `glibc` but
there surely are `libc` implementations that would be OK with such
a username within system, so your solution is not as absolute as
you may think, it's just based on obscurity at the situation there's no
need to hide anything.

Reply:
The situation addressed here is whenever getpwuid() fails. When that is the case, we cannot enforce ACLs, but that does not prevent non-ACL operations from working. Previously, the daemons could assert in this situation, and it was possible to get to a point where the cluster did not detect and recover from resource failures. With this fix, the daemons no longer assert, only ACL operations are affected, and the cluster is able to recover.

It doesn't matter whether users can set ACLs for the fictitious user used here. That's not a problem and doesn't require a solution. The fictitious user is just a simple way of staying out of the way of real ACLs. It's not based on obscurity, and people are welcome to go through acrobatics to set an ACL for it if they so desire, though I see no use case for that.

A complex solution as described above is neither warranted on time/benefit terms, nor an improvement. If getpwuid() is failing, a serious OS failure such as disk loss is far more likely the cause than SELinux suddenly preventing /etc/passwd access, and attempting to parse the log or manipulate the CIB for maintenance mode etc. is likely to fail.

Our goal is to address problems faced by real users in real-world situations. There are a practically infinite number of theoretically possible scenarios that would require an infinite amount of developer time to address, so we have to triage based on real-world impact. Any SELinux changes are far more likely to occur at cluster setup time, or in a development environment, or during a maintenance window, where the user can easily identify and address problems that arise, than during normal cluster operation, when hardware failures are a prime scenario that HA is meant to address.

Reply:
The _obscurity_ I am talking about is in the fact you intentionally
code this new artificial user so that it cannot have the system-wide
counterpart for most of the time, except that it is not a universal
truth and other libc and systems may be more relaxed, so they would
allow havine `#unprivileged` user defined just fine.

And my question is then why to bother to try to do any kind of
hiding, when it can be __PCMK-UNPRIVILEGED-USER__ just fine,
with the advantage any such _acrobatics_ how you call it (it could
also be a stable workaround for remote only user database when
the connection may intermittently hang) would survive imminent
new lexical restriction on `acl_target/@id` (per `NCName` datatype),
for it to be ID-unique-enforcable with schema, which it currently
is not.

Reply:
Also note that any risks and shortcomings there are put solely on the users,
they would need to define such a user in ACLs in the first place (as an
enabling precondition), so I don't think there's anything against using
a non-obscured (generally system-wide usable) yet 1:billions+ improbably
clashing system-wide user.  Would you accept such a change?

Reply:
I don't understand the use case. Why would someone want an ACL for unidentifiable users during a system crisis? This is simply a less brittle method of defaulting to no privileges in an error situation than coding "if user is NULL" at every operation that requires an ACL. To me, having the schema make it impossible to set an ACL for this user is a bonus, if anything.